### PR TITLE
Global IOC Helpers

### PR DESCRIPTION
**Pull request by wey-chiang**
### Background

A few simple rules to look for activity from known IOCs related to the sunburst compromise.

### Changes

* Created a new global helper: `panther_iocs` : `panther_iocs.py` 
* Created a new folder for IOC rules: `panther_ioc_rules`
* Added Sunburst IP/FQDN/SHA-256 IOC rules

### Testing

* `make test`

### CODE REVIEW

File: panther_ioc_rules/sunburst_fqdn_iocs.py
Comment (nhakmiller) [inline panther_ioc_rules/sunburst_fqdn_iocs.py:9]: Let's gussy this up a bit by adding at least the `p_any_domains` field to this. Ideally, we could specify the exact FQDN that matched. You could take the intersection off the `p_any_domains` field and the ioc's field to see which domain it was exactly (maybe with some escaping e.g. replace `.`  with `[.]` so that someone doesn't actually click a known malicious domain).

Same comment for the other rules.

File: panther_ioc_rules/sunburst_fqdn_iocs.yml
Comment (nhakmiller) [inline panther_ioc_rules/sunburst_fqdn_iocs.yml:33]: I'd add "Consider rotating credentials on any systems observed communicating with these known malicious systems" or something similar.

File: global_helpers/panther_iocs.py
Comment (nhakmiller) [inline global_helpers/panther_iocs.py:47]: I almost feel like this should just be in the rule,  and we should just let `SUNBURST_FQDN_IOCS` be in the global. I could go either way though.

**Review by nhakmiller — COMMENTED** [review]
Lets update the title functions a bit then I think this looks good.

File: global_helpers/panther_iocs.py
Comment (wey-chiang) [reply global_helpers/panther_iocs.py:51]: I'm on the fence, could go either way -- I like the idea of having the `panther_iocs.py` file just containing any helper functions like `intersection` and just the IOCs.

Comment (wey-chiang) [reply global_helpers/panther_iocs.py:51]: Should probably move the string replace functions into their own function in `panther_iocs.py`, 

e.g.

```
def sanitize_title(title: str) -> str:
    single_quote = '\''
    return title.replace('.', '[.]').replace(single_quote, '')
```

File: global_helpers/panther_iocs.py
Comment (nhakmiller) [inline global_helpers/panther_iocs.py:63]: Let's drop this, unless there are specific instances you have in mind where this covers something the `p_any_ip_addresses` doesn't cover.

File: panther_ioc_rules/sunburst_fqdn_iocs.py
Comment (nhakmiller) [inline panther_ioc_rules/sunburst_fqdn_iocs.py:10]: At this point we know that `p_any_domain_names` must be in `event`, otherwise we'd never have gotten to this point right? I'd remove this `if` statement, I think it might be confusing about because it might lead you to wonder how you could get here if `p_any_domain_names` is not present.

File: global_helpers/panther_iocs.py
Comment (nhakmiller) [reply global_helpers/panther_iocs.py:51]: I like the `sanitize_title` idea. I think how you have it now is good, we can leave it like this to get this first iteration out the door then consider iterating in the future.

File: global_helpers/panther_iocs.py
Comment (jacknagz) [inline global_helpers/panther_iocs.py:3]: Shouldn't we just use sets for all of these IOCs if we are checking for existence? cc @austinbyers

File: global_helpers/panther_iocs.py
Comment (jacknagz) [inline global_helpers/panther_iocs.py:40]: I would generalize these three functions to just take in the event + a set of indicators. Otherwise, you need a new function for every indicator list

File: global_helpers/panther_iocs.py
Comment (jacknagz) [reply global_helpers/panther_iocs.py:63]: Better yet - sub this for the UDM calls

File: global_helpers/panther_iocs.yml
Comment (jacknagz) [inline global_helpers/panther_iocs.yml:4]:
```suggestion
Description: Global helpers and variables used for finding hits on public indicators of compromise (IOCs) 
```

File: panther_ioc_rules/sunburst_fqdn_iocs.py
Comment (jacknagz) [inline panther_ioc_rules/sunburst_fqdn_iocs.py:1]: Should we have separate rules for IoC matches? Or a single rule that operates against multiple lists of IoCs?

**Review by jacknagz — CHANGES_REQUESTED** [review]
Great start!

File: global_helpers/panther_iocs.py
Comment (nhakmiller) [reply global_helpers/panther_iocs.py:63]: No UDM's have been published yet, pending this PR: https://github.com/panther-labs/panther-analysis/pull/156. We can update with UDM soon though.

File: panther_ioc_rules/sunburst_fqdn_iocs.py
Comment (nhakmiller) [reply panther_ioc_rules/sunburst_fqdn_iocs.py:1]: Good question, I could go either way on it to honest. I do kind of like separate rules as it gives you a little bit more knobs to turn with regards to which log types apply which logic, but it's not a big deal. I'm pushing up a change with three rules, but if you guys decide to reduce it to one it should be pretty easy.

File: global_helpers/panther_iocs.py
Comment (nhakmiller) [reply global_helpers/panther_iocs.py]: I ended up compromising, I left a very generic version called `ioc_match` in the global and put a bit more heavy lifting into the rule itself. I think this strikes a good balance avoiding duplicated code, while keeping the different rules separate.

**Review by nhakmiller — APPROVED** [review]
Looks good to me, but I've re-written some of it so I'll leave it to another reviewer for final approval + merge.

@lindsey-w @wey-chiang , since I'll be out tomorrow if someone could re-review (and make any changes needed), merge, and then cut a new release tomorrow morning following the process we discussed previously that would be very great.

File: panther_ioc_rules/sunburst_fqdn_iocs.py
Comment (wey-chiang) [reply panther_ioc_rules/sunburst_fqdn_iocs.py]: > Should we have separate rules for IoC matches? Or a single rule that operates against multiple lists of IoCs?

Single rule that operates against multiple lists of IoCs IMO

We can get that same benefit @nhakmiller by making a rule for each indicator type that we match on i.e. fqdn, ip, sha256

File: panther_ioc_rules/sunburst_fqdn_iocs.yml
Comment (jacknagz) [inline panther_ioc_rules/sunburst_fqdn_iocs.yml:31]: Monitors for communication to known Sunburst Backdoor FQDNs

File: panther_ioc_rules/sunburst_fqdn_iocs.yml
Comment (jacknagz) [inline panther_ioc_rules/sunburst_fqdn_iocs.yml:32]: Let's add a reference:

https://www.fireeye.com/blog/threat-research/2020/12/evasive-attacker-leverages-solarwinds-supply-chain-compromises-with-sunburst-backdoor.html

File: panther_ioc_rules/sunburst_fqdn_iocs.yml
Comment (jacknagz) [reply panther_ioc_rules/sunburst_fqdn_iocs.yml:31]: Let's use this formula for the description and replicate it to the other rules

**Review by jacknagz — APPROVED** [review]
LGTM after final edits to description/reference! 🚢

**Pull request merged**
